revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,235 @@
+// Package OCSP provides methods for checking the OCSP revocation status of a
+// certificate chain, as well as errors related to these checks
+package ocsp
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	coreX509 "github.com/notaryproject/notation-core-go/x509"
+	"golang.org/x/crypto/ocsp"
+)
+
+// Options specifies values that are needed to check OCSP revocation
+type Options struct {
+	CertChain   []*x509.Certificate
+	SigningTime time.Time
+	HTTPClient  *http.Client
+}
+
+const (
+	pkixNoCheckOID      string = "1.3.6.1.5.5.7.48.1.5"
+	ocspMaxResponseSize int64  = 20480 //bytes
+)
+
+// CheckStatus checks OCSP based on the passed options and returns a 2D array of
+// errors. The size of these dimensions is as follows:
+//   - 1st dimension will always be equal in length to the certificate chain.
+//   - 2nd will have a length between 1 and the number of OCSPServers for the cert
+//
+// At each index, nil will be used to indicate no error. If there is an error,
+// it will be one of the errors defined in errors.go.
+//
+// (e.g. for a chain with 3 certs (leaf, intermediate, root) with both the leaf
+// and intermediate specifying 2 OCSP server, this could be a possible result):
+//
+//	[[PKIXNoCheckError, TimeoutError], [RevokedError], [nil]]
+func CheckStatus(opts Options) [][]error {
+	result := make([][]error, len(opts.CertChain))
```

Comment:
We need 2D array because there is an array of certificates that each have multiple servers that could each have an error. If we don't merge the server responses, then a 2D array is necessary.

If this is asking why we aren't using an array of CertRevocationResult objects for ocsp.CheckStatus, that is an implementation specifically for the revocation package. The raw errors from ocsp.CheckStatus are interpreted into this structure inside revocation. This will allow us to easily update and combine the results once we incorporate CRL later.